### ISSUE TITLE
Arb.register calls are ignored in parallel FsCheck tests

### ISSUE THREAD

**Issue by dungpa** [fscheck/FsCheck#448]
I have a number of FsCheck tests that run slowly.
I tried to upgrade to https://www.nuget.org/packages/FsCheck/3.0.0-alpha3 to take advantage of parallel execution.

However, all the custom generators and shrinkers registered via `Arb.register` are ignored whenever `ParallelRunConfig` is enabled.

I'm aware of `Prop.forAll` but it seems to be very limited. Our custom generators are for basic types (numbers, strings, datetimes, etc.) but the generated test inputs are big structural types, which are very cumbersome to write explicit generators for.

Is there any good way to register custom basic arbitraries for parallel execution?

**Comment by kurtschelfthout** [comment]
I think it is supposed to work. There is even a test for it. Do you have a repro?

**Comment by dungpa** [comment]
Thanks for the answer.

Here is an example:

```fsharp
    module Repro

    open NUnit.Framework
    open FsCheck

    type Generator =
        static member NonNullString () =
            Arb.Default.String ()
            |> Arb.filter(isNull>>not)

    type Input =
        | String of string
        | Int of int
    
    let prop = function
        | String s -> not (isNull s)
        | Int _ -> true

    [<Test>]
    let ``Can't handle null string``() =
        Arb.register<Generator>() |> ignore
        Check.One({ Config.QuickThrowOnFailure with 
                        MaxTest = 100; 
                        ParallelRunConfig = Some { MaxDegreeOfParallelism = System.Environment.ProcessorCount } }, prop)
```
Expected: test should pass

Actual: 

```
Result Message:	
System.Exception : Falsifiable, after 29 tests (0 shrinks) (7949183667445137684,9415070992166401363)
Last step was invoked with size of 30 and seed of (10950565245694646435,16673202306257149031):
Original:
String null
with exception:
System.Exception: Expected true, got false.
```

Environment: 

FsCheck 3.0.0-alpha3
NUnit 3.8
FSharp.Core 4.4.1.0 (via binding redirects)

**Comment by kurtschelfthout** [comment]
Thanks for the repro - haven't had time to look into yet but  I will.

**Comment by dungpa** [comment]
@kurtschelfthout Have you had a chance to have a look?

**Comment by kurtschelfthout** [comment]
Yes. You are right, it doesn't work, and looking at it the tests that were added to check that this works are naive (and must be false positives). The arbitrary instances that are registered are stored in a thread-local variable. This is a "legacy" from v2 where that mostly works in the sense that it allows FsCheck to work with xUnit parallel test runs (which apparently uses separate threads). But in v3 FsCheck's native paralellism uses threadpool threads itself, and so the thread-local doesn't get propagated.

I always knew and expected this to be a problem. I'm still not sure how to solve this. I do strongly empathize with your use case, see also #396.

For now perhaps I am thinking of a "quick fix" which is to explicitly set/unset the threadlocal when a new "task" is posted to the threadpool.

**Comment by dungpa** [comment]
It looks like a fine quick fix. For my use case, I often have a set of fixed arbitraries per test assembly. The main purpose is to filter out irrelevant values from test domains.

Parallel tests by NUnit/xUnit seems too coarse-grained when I have 10_000 test cases per test.

**Comment by ImaginaryDevelopment** [comment]
@dungpa can you show the updated code with the work-around?

**Comment by kurtschelfthout** [comment]
@ImaginaryDevelopment  there is no workaround, it needs a change to FsCheck. Not much you can do from  client code I think.

**Comment by ImaginaryDevelopment** [comment]
> For now perhaps I am thinking of a "quick fix" which is to explicitly set/unset the threadlocal when a new "task" is posted to the threadpool.

I guess the quick fix you were referring to was inside FsCheck, not the client code? Alright.
